File: app/src/main/java/de/westnordost/streetcomplete/data/osm/persist/OsmQuestTable.java
package de.westnordost.streetcomplete.data.osm.persist;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class OsmQuestTable
{
	public static final String NAME = "osm_quests";
	public static final String NAME_MERGED_VIEW = "osm_quests_full";

	public static final String NAME_UNDO = "osm_quests_undo";
	public static final String NAME_UNDO_MERGED_VIEW = "osm_quests_full_undo";


	public static class Columns
	{
		public static final String
				QUEST_ID = "quest_id",
				QUEST_TYPE = "quest_type",
				ELEMENT_ID = "element_id",
				ELEMENT_TYPE  = "element_type",
				QUEST_STATUS = "quest_status",
				TAG_CHANGES = "tag_changes",
				CHANGES_SOURCE = "changes_source",
				LAST_UPDATE = "last_update";

		public static final List<String> ALL = Collections.unmodifiableList(Arrays.asList(
				QUEST_ID, QUEST_TYPE, ELEMENT_ID, ELEMENT_TYPE, QUEST_STATUS, TAG_CHANGES,
				CHANGES_SOURCE, LAST_UPDATE));

		public static final List<String> ALL_DB_VERSION_3 = Collections.unmodifiableList(Arrays.asList(
				QUEST_ID, QUEST_TYPE, ELEMENT_ID, ELEMENT_TYPE, QUEST_STATUS, TAG_CHANGES,
				LAST_UPDATE));
	}
}

File: app/src/main/java/de/westnordost/streetcomplete/data/osm/persist/OsmQuestTable.kt
package de.westnordost.streetcomplete.data.osm.persist

object OsmQuestTable {
    const val NAME = "osm_quests"
    const val NAME_MERGED_VIEW = "osm_quests_full"

    object Columns {
        const val QUEST_ID = "quest_id"
        const val QUEST_TYPE = "quest_type"
        const val ELEMENT_ID = "element_id"
        const val ELEMENT_TYPE = "element_type"
        const val QUEST_STATUS = "quest_status"
        const val TAG_CHANGES = "tag_changes"
        const val CHANGES_SOURCE = "changes_source"
        const val LAST_UPDATE = "last_update"
    }

    const val CREATE = """
        CREATE TABLE $NAME (
            ${Columns.QUEST_ID} INTEGER PRIMARY KEY,
            ${Columns.QUEST_TYPE} varchar(255) NOT NULL,
            ${Columns.QUEST_STATUS} varchar(255) NOT NULL,
            ${Columns.TAG_CHANGES} blob,
            ${Columns.CHANGES_SOURCE} varchar(255),
            ${Columns.LAST_UPDATE} int NOT NULL,
            ${Columns.ELEMENT_ID} int NOT NULL,
            ${Columns.ELEMENT_TYPE} varchar(255) NOT NULL,
            CONSTRAINT same_osm_quest UNIQUE (
                ${Columns.QUEST_TYPE},
                ${Columns.ELEMENT_ID},
                ${Columns.ELEMENT_TYPE}
            ),
            CONSTRAINT element_key FOREIGN KEY (
                ${Columns.ELEMENT_TYPE},
                ${Columns.ELEMENT_ID}
            ) REFERENCES ${ElementGeometryTable.NAME} (
                ${ElementGeometryTable.Columns.ELEMENT_TYPE},
                ${ElementGeometryTable.Columns.ELEMENT_ID}
            )
        );"""

    const val CREATE_VIEW = """
        CREATE VIEW $NAME_MERGED_VIEW
        AS SELECT * FROM $NAME
        INNER JOIN ${ElementGeometryTable.NAME}
        USING (
            ${ElementGeometryTable.Columns.ELEMENT_TYPE},
            ${ElementGeometryTable.Columns.ELEMENT_ID}
        );"""

    val ALL_COLUMNS_DB_VERSION_3 = listOf(
        Columns.QUEST_ID,
        Columns.QUEST_TYPE,
        Columns.ELEMENT_ID,
        Columns.ELEMENT_TYPE,
        Columns.QUEST_STATUS,
        Columns.TAG_CHANGES,
        Columns.LAST_UPDATE
    )

    const val CREATE_DB_VERSION_3 = """
        CREATE TABLE $NAME (
            ${Columns.QUEST_ID} INTEGER PRIMARY KEY,
            ${Columns.QUEST_TYPE} varchar(255) NOT NULL,
            ${Columns.QUEST_STATUS} varchar(255) NOT NULL,
            ${Columns.TAG_CHANGES} blob,
            ${Columns.LAST_UPDATE} int NOT NULL,
            ${Columns.ELEMENT_ID} int NOT NULL,
            ${Columns.ELEMENT_TYPE} varchar(255) NOT NULL,
            CONSTRAINT same_osm_quest UNIQUE (
                ${Columns.QUEST_TYPE},
                ${Columns.ELEMENT_ID},
                ${Columns.ELEMENT_TYPE}
            ),
            CONSTRAINT element_key FOREIGN KEY (
                ${Columns.ELEMENT_TYPE}, ${Columns.ELEMENT_ID}
            ) REFERENCES ${ElementGeometryTable.NAME} (
                ${ElementGeometryTable.Columns.ELEMENT_TYPE},
                ${ElementGeometryTable.Columns.ELEMENT_ID}
            )
        );"""
}
